packages/flutter/lib/src/foundation/persistent_hash_map.dart

```diff
@@ -0,0 +1,340 @@
+    if (newroot == _root) {
+      return this;
+    }
+    return PersistentHashMap<K, V>._(newroot);
+  }
+
+  /// Returns value associated with the given [key] or `null` if [key]
+  /// is not in the map.
+  V? lookup(K key) {
+    return _root != null ? _root!.lookup(0, key, key.hashCode) as V? : null;
+  }
+}
+
+/// Base class for nodes in a hash trie.
+abstract class _TrieNode {
+  static const int hashBitsPerLevel = 5;
+  static const int hashBitsPerLevelMask = (1 << hashBitsPerLevel) - 1;
+
+  @pragma('vm:prefer-inline')
+  static int trieIndex(int hash, int bitIndex) {
+    return (hash >>> bitIndex) & hashBitsPerLevelMask;
+  }
+
+  _TrieNode copyWith(int bitIndex, Object key, int keyHash, Object? value);
+
+  Object? lookup(int bitIndex, Object key, int keyHash);
+}
+
+/// A full (uncompressed) node in the trie.
+///
+/// It contains an array with `1<<_hashBitsPerLevel` elements which
+/// are references to deeper nodes.
+class _FullNode extends _TrieNode {
+  _FullNode(this.descendants);
+
+  static const int numElements = 1 << _TrieNode.hashBitsPerLevel;
+
+  final List<Object?> descendants;
+
+  @override
+  _TrieNode copyWith(int bitIndex, Object key, int keyHash, Object? value) {
+    final int index = _TrieNode.trieIndex(keyHash, bitIndex);
+    final _TrieNode node =
+        (descendants[index] as _TrieNode?) ?? _CompressedNode.empty;
+    final _TrieNode newNode = node.copyWith(
+        bitIndex + _TrieNode.hashBitsPerLevel, key, keyHash, value);
+    return identical(newNode, node)
+        ? this
+        : _FullNode(_copy(descendants)..[index] = newNode);
+  }
+
+  @override
+  Object? lookup(int bitIndex, Object key, int keyHash) {
+    final int index = _TrieNode.trieIndex(keyHash, bitIndex);
+    final _TrieNode? node = descendants[index] as _TrieNode?;
+    return node?.lookup(bitIndex + _TrieNode.hashBitsPerLevel, key, keyHash);
+  }
+}
+
+/// Compressed node in the trie.
+///
+/// Instead of storing the full array of outgoing edges this node uses a
+/// compressed representation:
+///
+///   * [_CompressedNode.occupied] has a bit set for indices which are occupied.
+///   * furthermore, each occupied index can either be a `(key, value)` pair
+///     representing an actual key/value mapping or a `(null, trieNode)` pair
+///     representing a descendant node.
+///
+class _CompressedNode extends _TrieNode {
+  _CompressedNode(this.occupiedIndices, this.keyValuePairs);
+  _CompressedNode._empty() : this(0, _emptyArray);
+
+  factory _CompressedNode.single(int bitIndex, int keyHash, _TrieNode node) {
+    final int bit = 1 << _TrieNode.trieIndex(keyHash, bitIndex);
+    // A single (null, node) pair.
+    final List<Object?> keyValuePairs = List<Object?>.filled(2, null)
+      ..[1] = node;
+    return _CompressedNode(bit, keyValuePairs);
+  }
+
+  static final _CompressedNode empty = _CompressedNode._empty();
+  static final List<Object?> _emptyArray = List<Object?>.filled(0, null);
```

Comment:
Yeah and this will make things slower. Rule of thumb: never mix growable and fixed arrays in the same field or variable - these arrays have different representation, so this introduces polymorphism which makes generated code worse. 

I will add a comment.